traitlets/traitlets.py

```diff
@@ -1124,11 +1125,6 @@ def hold(change):
                 self.notify_change = notify_change
                 self._cross_validation_lock = False
 
-                if isinstance(notify_change, types.MethodType):
-                    # Presence of the notify_change method
-                    # on __dict__ can cause memory leaks
```

Comment:
@minrk sorry about the delay, I've been pretty swamped recently.

This was almost a year ago, but I'm pretty sure this originates with `hold_trait_notifications` where `HasTraits.notify_change` is set and restored:

```python
_ = self.notify_change

def hold_(*args):
	# ...
self.notify_change = hold

yield

self.notify_change = _
```

So you end up getting a circular reference.

However looking at it now, `__dict__.pop` isn't needed if we do this instead:

```python
self.notify_change = hold

yield

del self.notify_change
```

This will also resolve pickling.